### PR TITLE
Improve web interface

### PR DESCRIPTION
**Pull request by TylerVigario**
Improvements for the web interface

- Many improvements to mobile layout and styling
- Moved javascript to `static/js/custom.js`

### CODE REVIEW

**Comment by TerryGeng** [comment]
Yeah you moved js into a file... I was too lazy and didn't do that.

File: templates/index.html [hunk 1]
```diff
@@ -145,263 +150,287 @@ <h1><img src="static/image/logo.png" style="width: 200px;" /> botamusique Web In
...
-                        </div>
-
-                        <div class="card-body">
-                            <div class="alert alert-secondary">
-                                <h4 class="alert-heading">Filters</h4>
-                                <div class="row">
-                                    <div class="col-6">
-                                        <div id="filter-type" class="form-group row">
-                                            <label class="col-sm-2 col-form-label">Type</label>
-                                            <div class="btn-group btn-group-toggle" data-toggle="buttons" style="height: 35px; padding-top:5px;">
```
Comment (TerryGeng) [inline templates/index.html:180]: col-sm-8 for this div?

**Comment by TylerVigario** [comment]
@TerryGeng Yes, absolutely. I'll see what improvements I can come up with.

**Comment by TylerVigario** [comment]
Okay, I committed some improvements to those areas. I got stuck on getting the filter section inner columns to collapse. I'll take another crack at it later.

File: templates/index.html [hunk 2]
```diff
@@ -6,17 +6,17 @@
 
     <title>botamusique web interface</title>
 
-    <link rel="icon" href="static/image/favicon.ico" />
+    <link rel="icon" href="../static/image/favicon.ico" />
```
Comment (TerryGeng) [inline templates/index.html:9]: Oh, are you sure this work?

Comment (TylerVigario) [reply templates/index.html:9]: Lol, I've been just editing the static file directly. How does your normal development setup look like for this project?

Comment (TerryGeng) [reply templates/index.html]: I just uncommemt L65 of interface.py then you can just refresh the and see your changes on the fly ;)

Comment (TerryGeng) [reply templates/index.html]: Oh you mean you haven't get the bot running.... Then I have no good idea...

**Comment by TylerVigario** [comment]
I wonder how open are the owners in adding a Node build system?

File: templates/index.html (hunk 2 above)
Comment (TylerVigario) [reply templates/index.html]: > Oh you mean you haven't get the bot running.... Then I have no good idea...

Yeah, at least not on my (default) Windows rig. I run it on my CentOS box that hosts my Murmur server though.

I'll keep the template auto-refresh in mind :see_no_evil:

**Comment by TerryGeng** [comment]
> I wonder how open the owners are to adding a Node build system?

Personally, I'm not a node developer and know little about it (actually I'm not even a software developer at all lol). I think this project is not a node project and am wondering how utilities of node can aid this project.

I think @azlux have the final say on these kind of topics and we can sound him out.

**Comment by TylerVigario** [comment]
> I think this project is not a node project and am wondering how utilities of node can aid this project.

You are absolutely right! It's a Python module/project and I have never touched Python.

However, it's a matter of tooling. And simply all the web assets are typically hosted on NPM.

**Comment by TerryGeng** [comment]
Hi, I've just tested again. Looks great! The only thing I noticed is the volume slider became too compressed. I think just omit it in the mobile view would be great, just as what you did to the thumbnail image. Would you consider this?

**Comment by TylerVigario** [comment]
Yeah, unfortunately, there are too many buttons on the top of the playlist section.

My current ideas are to

- Combine Play & Pause
- Volume slide vertically in hover when clicked
- Remove Stop button
- Simplify/reduce other buttons (One-shot, autoplay, etc.)

That should help out a bunch with that area. Let me know what you think.

**Comment by TerryGeng** [comment]
Thanks for your ideas!

> * Combine Play & Pause

Yes, I agree with you.

> * Volume slide vertically in hover when clicked

By doing so, we can actually remove the volume +/- button. I think this will work now since we have this nice volume slider from #129. We may consider using [popovers in bootstrap](https://getbootstrap.com/docs/4.0/components/popovers/) but it also means popper.min.js should be included.

> * Remove Stop button (possibly move elsewhere)

I'm thinking about turning this button into a fast-forward button since I usually use it to fast-forward. Stop isn't a necessity since pause was implemented #71. Its special role in one-shot mode can be done by the Clear Playlist button on the button. If you and @azlux  agree with me, I can start to work on the logic of it immediately.

> * Simplify/reduce other buttons (one-shot, random, repeat, and autoplay)

Bootstrap provides useful [dropdowns](https://getbootstrap.com/docs/4.0/components/dropdowns/). Do you think moving these playlist options into a dropdown will work for mobile users?

**Comment by azlux** [comment]
Should I merge or wait for futur commits ?

**Comment by TylerVigario** [comment]
If you don't mind giving it a few days I could find some time to implement the changes we outlined above. Otherwise, I can just send it in via another PR later.

**Comment by TerryGeng** [comment]
I wrote an action called *next*. If you'd like to implement the fast-forward button, this would be useful!

**Comment by TerryGeng** [comment]
Hey! It has been a few days since the last commit. Are you getting stuck in some places? Is there anything we can help with?

**Comment by TylerVigario** [comment]
Sorry, for the delay, I've been busy looking after my two little ones. I was able to get the play/pause logic mostly finished which I'll commit shortly.

**Comment by TylerVigario** [comment]
I've been working on the button dropdown for the play mode as well. It's nearly finished but I was having trouble in testing where it wouldn't change the play mode even though I feel it should have.

**Comment by TerryGeng** [comment]
image

Hi! Thanks for your work! Everything looks neat now! However, I would suggest put the icon of the current mode instead of a "Mode" on that dropdown button. And, personally I prefer this menu sits on the right-hand side, where it used to be. What's your opinion?

We have made some other modifications in the master branch, which makes it harder to test your branch (since the logic of the database part has changed and some dependencies problem). I think I can perform a merge as soon as you have fixed that little bug I reported so you can continue to work on the newest version of the bot.

**Comment by TylerVigario** [comment]
Should I merge the changes into my feature branch? Or just leave the merge strategies to you guys?

File: templates/index.html [hunk 3]
```diff
@@ -27,33 +27,37 @@ <h1>botamusique Web Interface</h1>
...
+                        <button class="btn btn-secondary dropdown-toggle" type="button" id="play-mode"
+                            data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                            Mode
+                        </button>
+                        <div class="dropdown-menu" aria-labelledby="play-mode">
+                            <a class="dropdown-item" href="#" id="one-shot-mode-btn"
+                                onclick="changePlayMode('one-shot')">
+                                <i class="fas fa-tasks mr-2" aria-hidden="true"></i>One-shot
+                            </a>
+                            <a class="dropdown-item" href="#" id="random-mode-btn" onclick="changePlayMode('random')">
```
[the remark this answers is not in the document]
Comment (TylerVigario) [reply templates/index.html:46]: Nice catch! I'll fix that now.

**Comment by TerryGeng** [comment]
By now it seems that auto-merging still works, but I'm just wondering if we keep working in a separate branch, someday we will face a merging problem.

Anyway, after I merge this branch, you can pull the master again and work on it :).

**Comment by TerryGeng** [comment]
A little reminder: if you want to run the bot after this update, please remember to run `python3 -m pip install pymumble` in your venv. Since we have left pymumble to pip, it becomes a new dependency.

**Comment by TylerVigario** [comment]
Sweet, will do!

**Comment by TerryGeng** [comment]
I finished some changes I mentioned above. Now it's your turn again :)

**Pull request merged**